File: app/src/main/java/com/igordanilchik/daggertest/HttpClientModule.java
package com.igordanilchik.daggertest;

import android.util.Log;

import java.io.IOException;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import okio.Buffer;
import okio.BufferedSource;
import okio.ForwardingSource;
import okio.Okio;
import okio.Source;

@Module
public class HttpClientModule {
    private static final String LOG_TAG = HttpClientModule.class.getSimpleName();

    @Provides
    @Singleton
    OkHttpClient provideHttpClient() {
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();

        if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
            logging.setLevel(HttpLoggingInterceptor.Level.HEADERS);
            httpClient.addInterceptor(logging);

            httpClient.addNetworkInterceptor(chain -> {
                okhttp3.Response originalResponse = chain.proceed(chain.request());
                return originalResponse.newBuilder()
                        .body(new ProgressResponseBody(originalResponse.body(), progressListener))
                        .build();
            });
        }
        return httpClient.build();
    }

    private static class ProgressResponseBody extends ResponseBody {

        private final ResponseBody responseBody;
        private final ProgressListener progressListener;
        private BufferedSource bufferedSource;

        ProgressResponseBody(ResponseBody responseBody, ProgressListener progressListener) {
            this.responseBody = responseBody;
            this.progressListener = progressListener;
        }

        @Override public MediaType contentType() {
            return responseBody.contentType();
        }

        @Override public long contentLength() {
            return responseBody.contentLength();
        }

        @Override public BufferedSource source() {
            if (bufferedSource == null) {
                bufferedSource = Okio.buffer(source(responseBody.source()));
            }
            return bufferedSource;
        }

        private Source source(Source source) {
            return new ForwardingSource(source) {
                long totalBytesRead = 0L;

                @Override public long read(Buffer sink, long byteCount) throws IOException {
                    long bytesRead = super.read(sink, byteCount);
                    // read() returns the number of bytes read, or -1 if this source is exhausted.
                    totalBytesRead += bytesRead != -1 ? bytesRead : 0;
                    if (responseBody.contentLength() > 0) {
                        progressListener.update(totalBytesRead, responseBody.contentLength(), bytesRead == -1);
                    } else {
                        progressListener.update(totalBytesRead, bytesRead == -1);
                    }
                    return bytesRead;
                }
            };
        }
    }

    interface ProgressListener {
        //Known content length
        void update(long bytesRead, long contentLength, boolean done);
        //Unknown content length
        void update(long bytesRead, boolean done);
    }

    private final ProgressListener progressListener = new ProgressListener() {
        @Override
        public void update(long bytesRead, long contentLength, boolean done) {
            update(bytesRead, done);
            Log.d(LOG_TAG, "Content length: " + contentLength);
            Log.d(LOG_TAG, String.format("%d%% done\n", (100 * bytesRead) / contentLength));
        }

        @Override
        public void update(long bytesRead, boolean done) {
            Log.d(LOG_TAG, "Bytes read: " + bytesRead);
            Log.d(LOG_TAG, "Done: " + done);
        }
    };
}

File: app/src/main/java/com/igordanilchik/rxjava2test/common/di/HttpClientModule.kt
package com.igordanilchik.rxjava2test.common.di

import com.igordanilchik.rxjava2test.BuildConfig
import dagger.Module
import dagger.Provides
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import okhttp3.logging.HttpLoggingInterceptor
import okio.*
import timber.log.Timber
import java.io.IOException
import javax.inject.Singleton

@Module
class HttpClientModule {

    private val progressListener = object : ProgressListener {

        override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
            update(bytesRead, done)
            Timber.d("Content length: %s", contentLength)
            Timber.d("%d%% done\n", 100 * bytesRead / contentLength)
        }

        override fun update(bytesRead: Long, done: Boolean) {
            Timber.d("Bytes read: %s", bytesRead)
            Timber.d("Done: %s", done)
        }
    }

    @Provides
    @Singleton
    internal fun provideHttpClient(): OkHttpClient {
        val httpClient = OkHttpClient.Builder()

        if (BuildConfig.DEBUG) {
            val logging = HttpLoggingInterceptor()
            logging.level = HttpLoggingInterceptor.Level.HEADERS
            httpClient.addInterceptor(logging)

            httpClient.addNetworkInterceptor { chain ->
                val originalResponse = chain.proceed(chain.request())
                originalResponse.newBuilder()
                        .body(originalResponse.body()?.let { ProgressResponseBody(it, progressListener) })
                        .build()
            }
        }
        return httpClient.build()
    }

    private class ProgressResponseBody internal constructor(
            private val responseBody: ResponseBody,
            private val progressListener: ProgressListener
    ) : ResponseBody() {

        private val bufferedSource: BufferedSource by lazy { Okio.buffer(source(responseBody.source())) }

        override fun contentType(): MediaType? = responseBody.contentType()

        override fun contentLength(): Long = responseBody.contentLength()

        override fun source(): BufferedSource = bufferedSource

        private fun source(source: Source): Source {
            return object : ForwardingSource(source) {
                var totalBytesRead = 0L

                @Throws(IOException::class)
                override fun read(sink: Buffer, byteCount: Long): Long {
                    val bytesRead = super.read(sink, byteCount)
                    // read() returns the number of bytes read, or -1 if this source is exhausted.
                    totalBytesRead += if (bytesRead != -1L) bytesRead else 0L
                    if (responseBody.contentLength() > 0) {
                        progressListener.update(totalBytesRead, responseBody.contentLength(), bytesRead == -1L)
                    } else {
                        progressListener.update(totalBytesRead, bytesRead == -1L)
                    }
                    return bytesRead
                }
            }
        }
    }

    internal interface ProgressListener {
        //Known content length
        fun update(bytesRead: Long, contentLength: Long, done: Boolean)

        //Unknown content length
        fun update(bytesRead: Long, done: Boolean)
    }
}
